pkg/i2gw/aggregator.go

```diff
@@ -284,34 +295,56 @@ func (rg *ingressRuleGroup) toHTTPRoute() (gatewayv1beta1.HTTPRoute, []error) {
 			Matches: []gatewayv1beta1.HTTPRouteMatch{*match},
 		}
 
-		var numWeightedBackends, totalWeightSet int32
-		for _, path := range paths {
-			backendRef, err := toBackendRef(path.path.Backend)
-			if err != nil {
-				errors = append(errors, err)
-				continue
-			}
-			if path.extra != nil && path.extra.canary != nil && path.extra.canary.weight != 0 {
-				weight := int32(path.extra.canary.weight)
-				backendRef.Weight = &weight
-				totalWeightSet += weight
-				numWeightedBackends++
+		backendRefs, errs := rg.calculateBackendRefWeight(paths)
+		errors = append(errors, errs...)
+		hrRule.BackendRefs = backendRefs
+
+		httpRoute.Spec.Rules = append(httpRoute.Spec.Rules, hrRule)
+	}
+
+	return httpRoute, errors
+}
+
+func (rg *ingressRuleGroup) calculateBackendRefWeight(paths []ingressPath) ([]gatewayv1beta1.HTTPBackendRef, field.ErrorList) {
+	var errors field.ErrorList
+	var backendRefs []gatewayv1beta1.HTTPBackendRef
+
+	var numWeightedBackends, totalWeightSet int32
+	var weightTotal = 100
+	for i, path := range paths {
+		backendRef, err := toBackendRef(path.path.Backend, field.NewPath("paths", "backends").Index(i))
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		if path.extra != nil && path.extra.canary != nil && path.extra.canary.weight != 0 {
+			weight := int32(path.extra.canary.weight)
+			backendRef.Weight = &weight
+			totalWeightSet += weight
+			numWeightedBackends++
+			if path.extra.canary.weightTotal > 0 {
+				weightTotal = path.extra.canary.weightTotal
 			}
-			hrRule.BackendRefs = append(hrRule.BackendRefs, gatewayv1beta1.HTTPBackendRef{BackendRef: *backendRef})
 		}
-		if numWeightedBackends > 0 && numWeightedBackends < int32(len(hrRule.BackendRefs)) {
-			weightToSet := (int32(100) - totalWeightSet) / (int32(len(hrRule.BackendRefs)) - numWeightedBackends)
-			for i, br := range hrRule.BackendRefs {
-				if br.Weight == nil {
-					br.Weight = &weightToSet
-					hrRule.BackendRefs[i] = br
-				}
+		backendRefs = append(backendRefs, gatewayv1beta1.HTTPBackendRef{BackendRef: *backendRef})
+	}
+	if numWeightedBackends > 0 && numWeightedBackends < int32(len(backendRefs)) {
+		weightToSet := (int32(weightTotal) - totalWeightSet) / (int32(len(backendRefs)) - numWeightedBackends)
+		if weightToSet < 0 {
+			weightToSet = 0
+		}
+		for i := range backendRefs {
+			if backendRefs[i].Weight == nil {
+				backendRefs[i].Weight = &weightToSet
+			}
+			if *backendRefs[i].Weight > int32(weightTotal) {
+				backendRefs[i].Weight = pointer.Int32(int32(weightTotal))
```

Comment:
not sure I understand the logic.

If the weight set is higher than weightTotal you set it to `weightTotal` ?

Reply:
Yes. This is performed in order to prevent the following example (which could happen until now):
You set an ingress directing to a canary deployment with a weight of 130 and a default weight-total of 100. The primary ingress weight will be -30.

Reply:
Maybe I misunderstand something, but what happens if you have two ingress, one with `120` and the other one with `120` ? (`weightTotal` is not set - means it is 100 by default)

Reply:
Answering myself - they will both get a 100 weight so it will be 50:50 in the traffic.

Now, what happens when one is `150` and the other one is `100`? again no `weightTotal` is set

Reply:
have a link?

Reply:
I haven't found a statement saying they do not support more than one, but this is what I did:

I followed https://kubernetes.github.io/ingress-nginx/examples/canary/
The traffic split when as expected.

Then I added an additional canary ingress.
Even when setting its weight to 100/100 and the rest of the ingresses to 0/100, the new one would not receive requests. The controller itself didn't have any errors though. It was weird.

Reply:
The traffic cannot be segmented as expected due to the logic here.  https://github.com/kubernetes/ingress-nginx/blob/5b35651a1a40f5506d06540c1fcd103dc41a16d5/rootfs/etc/nginx/lua/balancer.lua#L258-L266

Reply:
Well, I am a bit confused by the logic but it looks like like you are right @levikobi . However IIUC it is one canary ingress per ingress rule not per deployment (check the [Known Limitations section](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#canary))

How the code handles a situation of more than one ingress for the same host with Canary annotations? I think we don't account for this case however I do not think it is very important now as we should deprecate this logic soon and move to a more generic way for provider-specific annotations.

Reply:
> How the code handles a situation of more than one ingress for the same host with Canary annotations?

I don't think there can be a situation of more than one ingress for the same host with Canary annotations with the ingress-nginx implementation (I tried it out and it just didn't work)... But I may be wrong

> I do not think it is very important now as we should deprecate this logic soon and move to a more generic way for provider-specific annotations.

Agree